### PR TITLE
Add Thiea Manager with basic functionality

### PR DESCRIPTION
**Pull request by wsquan171**
This change adds the skeleton of Theia manager, including APIServer and a simple `NetworkPolicyRecommendation` CR controller. 

The code is structured following Antrea controller, and summarized below:
- `pkg/apis/crd` structured CRD definitions for k8s client generation (consumed by controllers)
- `pkg/apis/intelligence` structured REST payload definitions by API group and endpoints
- `pkg/apiserver/registry` REST handlers grouped  by API group and endpoints
- `pkg/apiserver/apiserver.go` APIServer definitions and configurations
- `pkg/client` k8s client-gen'ed codes
- `pkg/controller` collection of k8s resource controllers

This change also adds Helm chart generation and code generation scripts. 

To get the manager up and running in an existing testbed:
1. Apply `build/charts/theia/crds/network-policy-recommendation-crd.yaml`
2. `make theia-manager` and scp image `projects.registry.vmware.com/antrea/theia-manager:latest` to worker nodes
3. `hack/generate-manifest.sh --mode dev --theia-manager  > build/yamls/flow-visibility.yml` and apply the diffs

With the following CR deployed:
```
apiVersion: crd.theia.antrea.io/v1alpha1
kind: NetworkPolicyRecommendation
metadata:
  name: pr-test
spec:
  jobType: "initial"
  limit: 1000
  policyType: "anp-deny-all"
  startTime: 2022-08-24T10:45:57Z
```

`$ curl -k https://<theia-manager-svc>/apis/intelligence.theia.antrea.io/v1alpha1/networkpolicyrecommendations/pr-test`
```
{
  "kind": "NetworkPolicyRecommendation",
  "apiVersion": "intelligence.theia.antrea.io/v1alpha1",
  "metadata": {
    "name": "pr-test",
    "creationTimestamp": null
  },
  "limit": 1000,
  "policyType": "anp-deny-all",
  "startTime": "2022-08-24T10:45:57Z",
  "endTime": null
}
```

`curl -k https://<theia-manager-svc>apis/intelligence.theia.antrea.io/v1alpha1/networkpolicyrecommendations/pr-test-1`
```
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {},
  "status": "Failure",
  "message": "networkpolicyrecommendations.intelligence.theia.antrea.io \"pr-test-1\" not found",
  "reason": "NotFound",
  "details": {
    "name": "pr-test-1",
    "group": "intelligence.theia.antrea.io",
    "kind": "networkpolicyrecommendations"
  },
  "code": 404
}
```

### CODE REVIEW

File: build/charts/theia/README.md [hunk 1]
```diff
@@ -62,6 +62,12 @@ Kubernetes: `>= 1.16.0-0`
 | sparkOperator.enable | bool | `false` | Determine whether to install Spark Operator. It is required to run Network Policy Recommendation jobs. |
 | sparkOperator.image | object | `{"pullPolicy":"IfNotPresent","repository":"projects.registry.vmware.com/antrea/theia-spark-operator","tag":"v1beta2-1.3.3-3.1.1"}` | Container image used by Spark Operator. |
 | sparkOperator.name | string | `"policy-recommendation"` | Name of Spark Operator. |
+| theiaManager.apiServer.apiPort | int | `11347` | The port for the Flow Aggregator APIServer to serve on. |
```
Comment (yuntanghsu) [inline build/charts/theia/README.md:65]:
```suggestion
| theiaManager.apiServer.apiPort | int | `11347` | The port for the Theia Manager APIServer to serve on. |
```

File: build/charts/theia/conf/theia-manager.conf [hunk 2]
```diff
@@ -0,0 +1,13 @@
+# apiServer contains APIServer related configuration options.
+apiServer:
+  # The port for the flow-aggregator APIServer to serve on.
```
Comment (yuntanghsu) [inline build/charts/theia/conf/theia-manager.conf:3]:
```suggestion
  # The port for the theia-manager APIServer to serve on.
```

File: build/charts/theia/values.yaml [hunk 3]
```diff
@@ -198,3 +198,22 @@ sparkOperator:
...
+  # -- Determine whether to install Theia Manager.
+  enable: false
+  # -- Container image used by Theia Manager.
+  image:
+    repository: "projects.registry.vmware.com/antrea/theia-manager"
+    pullPolicy: "IfNotPresent"
+    tag: ""
+  # apiServer contains APIServer related configuration options.
+  apiServer:
+    # -- The port for the Flow Aggregator APIServer to serve on.
```
Comment (yuntanghsu) [inline build/charts/theia/values.yaml:211]:
```suggestion
    # -- The port for the Theia Manager APIServer to serve on.
```

File: pkg/config/theiamanager/config.go [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+package config
+
+type TheiaManagerConfig struct {
+	// apiServer contains APIServer related configuration options.
+	APIServer APIServerConfig `yaml:"apiServer,omitempty"`
+}
+
+type APIServerConfig struct {
+	// APIPort is the port for the antrea-agent APIServer to serve on.
+	// Defaults to 10352.
```
Comment (yuntanghsu) [inline pkg/config/theiamanager/config.go:24]:
```suggestion
	// APIPort is the port for the theia-manager APIServer to serve on.
	// Defaults to 11347.
```

File: build/charts/theia/README.md (hunk 1 above)
Comment (wsquan171) [reply build/charts/theia/README.md]: Thanks! addressed these typos when copy-pasting.

File: pkg/apiserver/apiserver.go [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+		return nil, err
+	}
+
+	// TODO (wshaoquan): use native k8s role-based auth
+	serverConfig.Authorization = genericapiserver.AuthorizationInfo{
+		Authorizer: authorizerfactory.NewAlwaysAllowAuthorizer()}
+	if err := os.MkdirAll(path.Dir(TokenPath), os.ModeDir); err != nil {
+		return nil, fmt.Errorf("error when creating dirs of token file: %v", err)
+	}
+	if err := ioutil.WriteFile(TokenPath, []byte(serverConfig.LoopbackClientConfig.BearerToken), 0600); err != nil {
```
Comment (yuntanghsu) [inline pkg/apiserver/apiserver.go:111]: Do we need to remove ioutil in this PR as well?

File: cmd/theia-manager/options.go [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+// validate validates all the required options.
+func (o *Options) validate(args []string) error {
+	if len(args) != 0 {
+		return errors.New("no positional arguments are supported")
+	}
+	return nil
+}
+
+func (o *Options) loadConfigFromFile(file string) (*managerconfig.TheiaManagerConfig, error) {
+	data, err := ioutil.ReadFile(file)
```
Comment (yuntanghsu) [inline cmd/theia-manager/options.go:68]: ditto

File: pkg/apiserver/apiserver.go (hunk 5 above)
Comment (wsquan171) [reply pkg/apiserver/apiserver.go]: done

**Review by yanjunz97 — COMMENTED** [review]
Thanks Shawn for working on this!

File: pkg/controller/networkpolicyrecommendation/controller.go [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+			klog.Errorf("Error decoding object tombstone when deleting NP Recommendation, invalid type: %v", tombstone.Obj)
+			return
+		}
+	}
+	klog.V(2).Infof("Processing NP Recommendation %s DELETE event, labels: %v", npReco.Name, npReco.Labels)
+	namespacedName := apimachinerytypes.NamespacedName{
+		Namespace: npReco.Namespace,
+		Name:      npReco.Name,
+	}
+	c.queue.Add(namespacedName)
```
Comment (yuntanghsu) [inline pkg/controller/networkpolicyrecommendation/controller.go:106]: Should it be c.queue.forget(namespacedName)?

File: build/images/Dockerfile.theia-manager.ubuntu [hunk 8]
```diff
@@ -0,0 +1,20 @@
+ARG GO_VERSION
+FROM golang:${GO_VERSION} as theia-manager-build
+
+COPY . /theia
+WORKDIR /theia
+
+# Statically links theia-manager binary.
+RUN CGO_ENABLED=0 make theia-manager-bin
```
[the remark this answers is not in the document]
Comment (wsquan171) [reply build/images/Dockerfile.theia-manager.ubuntu]: Thanks for pointing out. fixed

File: cmd/theia-manager/theia-manager.go [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+	// Set up signal capture: the first SIGTERM / SIGINT signal is handled gracefully and will
+	// cause the stopCh channel to be closed; if another signal is received before the program
+	// exits, we will force exit.
+	stopCh := signals.RegisterSignalHandlers()
+
+	log.StartLogFileNumberMonitor(stopCh)
+
+	kubeConfig, err := rest.InClusterConfig()
+	if err != nil {
+		return fmt.Errorf("error generating KubeConfig: %v", err)
```
[the remark this answers is not in the document]
Comment (wsquan171) [reply cmd/theia-manager/theia-manager.go]: done

File: build/charts/theia/crds/network-policy-recommendation-crd.yaml [hunk 10]
```diff
@@ -0,0 +1,72 @@
+apiVersion: apiextensions.k8s.io/v1
```
[the remark this answers is not in the document]
Comment (wsquan171) [reply build/charts/theia/crds/network-policy-recommendation-crd.yaml:1]: Sure. moved the common logic for `SPARK_OP` and `THEIA_MANAGER` under a new flag `INCLUDE_CRD`, and added new CRD to kustomize base if THEIA_MANAGER is set.

File: cmd/theia-manager/theia-manager.go [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+	"antrea.io/theia/pkg/controller/networkpolicyrecommendation"
+)
+
+// informerDefaultResync is the default resync period if a handler doesn't specify one.
+// Use the same default value as kube-controller-manager:
+// https://github.com/kubernetes/kubernetes/blob/release-1.17/pkg/controller/apis/config/v1alpha1/defaults.go#L120
+const informerDefaultResync = 12 * time.Hour
+
+func run(o *Options) error {
+	klog.Infof("Theia manager starting...")
```
Comment (dreamtalen) [inline cmd/theia-manager/theia-manager.go:39]: better to use structured logging: klog.InfoS(). Same for other logs.

File: pkg/apis/crd/v1alpha1/types.go [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+type NetworkPolicyRecommendation struct {
```
Comment (dreamtalen) [inline pkg/apis/crd/v1alpha1/types.go:29]: We also need to define the `RecommendedNetworkPolicy` CRD, but probably not necessary to include it in this PR.

File: pkg/apis/crd/v1alpha1/types.go [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+	StartTime           metav1.Time `json:"startTime,omitempty"`
+	EndTime             metav1.Time `json:"endTime,omitempty"`
+	NSAllowList         []string    `json:"nsAllowList,omitempty"`
+	ExcludeLabels       bool        `json:"excludeLabels,omitempty"`
+	ToServices          bool        `json:"toServices,omitempty"`
+	ExecutorInstances   int         `json:"executorInstances,omitempty"`
+	DriverCoreRequest   string      `json:"driverCoreRequest,omitempty"`
+	DriverMemory        string      `json:"driverMemory,omitempty"`
+	ExecutorCoreRequest string      `json:"executorCoreRequest,omitempty"`
+	ExecutorMemory      string      `json:"executorMemory,omitempty"`
```
Comment (dreamtalen) [inline pkg/apis/crd/v1alpha1/types.go:50]: We may need to include `wait` and `file` flags inside the spec struct.

File: pkg/apis/intelligence/v1alpha1/types.go [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+
+package v1alpha1
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+type JobType string
+
+const (
+	NPRecommendationJobInitial    JobType = "Initial"
+	NPRecommendationJobSubsequent JobType = "Subsequent"
```
Comment (dreamtalen) [inline pkg/apis/intelligence/v1alpha1/types.go:23]: May I ask why we create const for JobType? And why JobType only?

File: pkg/apiserver/apiserver.go [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+	NPRecommendationQuerier querier.NPRecommendationQuerier
+}
+
+func (s *theiaManagerAPIServer) Run(stopCh <-chan struct{}) error {
+	if s == nil {
+		return fmt.Errorf("nil s")
+	}
+	p := s.GenericAPIServer
+	if p == nil {
+		return fmt.Errorf("nil")
```
Comment (dreamtalen) [inline pkg/apiserver/apiserver.go:76]: ditto

File: pkg/apiserver/apiserver.go [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+}
+
+type theiaManagerAPIServer struct {
+	GenericAPIServer        *genericapiserver.GenericAPIServer
+	NPRecommendationQuerier querier.NPRecommendationQuerier
+}
+
+func (s *theiaManagerAPIServer) Run(stopCh <-chan struct{}) error {
+	if s == nil {
+		return fmt.Errorf("nil s")
```
Comment (dreamtalen) [inline pkg/apiserver/apiserver.go:72]: a single `s` may be confusing when debugging, better to add more info in this error message.

File: cmd/theia-manager/theia-manager.go (hunk 11 above)
Comment (wsquan171) [reply cmd/theia-manager/theia-manager.go]: fixed 4 occurrences

File: pkg/apis/crd/v1alpha1/types.go (hunk 12 above)
Comment (wsquan171) [reply pkg/apis/crd/v1alpha1/types.go:29]: yes. this controller was meant to demo basic introp between api server and controller, so not all crds are created.

File: pkg/apis/crd/v1alpha1/types.go (hunk 13 above)
Comment (wsquan171) [reply pkg/apis/crd/v1alpha1/types.go:50]: `wait` and `file` are more of a cli / client oriented spec. For CRD (abstraction of NP reco job), I don't think it makes much sense to include these 2? `wait` was to make cli watch for the result, while `file` defines a local file path to store the result on the caller side. Both doesn't seem to be applicable to this K8s resource.

File: pkg/apis/intelligence/v1alpha1/types.go (hunk 14 above)
Comment (wsquan171) [reply pkg/apis/intelligence/v1alpha1/types.go:23]: for testing and providing a template of where to define these const. more can be added when consumed.

File: pkg/apiserver/apiserver.go (hunk 15 above)
Comment (wsquan171) [reply pkg/apiserver/apiserver.go]: removed both. was leftover when doing local debugging. neither should be the case when actually running the code.

File: pkg/controller/networkpolicyrecommendation/controller.go (hunk 7 above)
Comment (wsquan171) [reply pkg/controller/networkpolicyrecommendation/controller.go:106]: `c.queue` is the actual worker queue to reconcile the resources. here we're just adding the key to the workqueue to indicate that some resource is removed and reconciliation is needed (i.e. deletion of child res related to this job CR). the key is forgotten only when reconciliation is done for it.

**Review by yanjunz97 — APPROVED** [review]
LGTM

**Review by yuntanghsu — APPROVED** [review]

**Review by salv-orlando — APPROVED** [review]
LGTM

**Pull request merged**